### PR TITLE
Homework5

### PR DESCRIPTION
**Pull request by Pokosteler**
java Beginner Chapter 2

### CODE REVIEW

File: src/lesson5/javaBeginnerChapter2/Task20.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+            for (int i = 0; i < str.length(); i++) {
+                if (str.charAt(i) == ' ') {
+                    newText += str.charAt(i - 1);
+                }
+                if (i == str.length() - 1) {
+                    newText += str.charAt(i);
+                }
+            }
+        }
+        System.out.println(newText);
```
Comment (TheFallen058) [inline src/lesson5/javaBeginnerChapter2/Task20.java:24]: You forgot about dot, comma and other symbols. They are not letters, you can't create a word from them

File: src/lesson5/javaBeginnerChapter2/Task21.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package lesson5.javaBeginnerChapter2;
+
+import java.util.Scanner;
+
+//Напишите три цикла выполняющих многократное сложение строк, один с
+//помощью оператора сложения и String, другой с помощью StringBuilder и метода
+//append, а также аналогино для StringBuffer. Сравните скорость их выполнения.
+public class Task21 {
+    public static void main(String[] args) {
+        Scanner sc = new Scanner(System.in);
```
Comment (TheFallen058) [inline src/lesson5/javaBeginnerChapter2/Task21.java:10]: Of course I expect that should enter text only once and get result from all cycles.

File: src/lesson5/javaBeginnerChapter2/Task21.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            for (int i = 0; i < str3.length(); i++) {
+                if (str3.charAt(i) != ' ') {
+                    newText3.append(str3.charAt(i));
+                }
+            }
+            long end = System.nanoTime();
+            System.out.println("String: " + newText3);
+            System.out.println("Total execution time: " + (end - start));
+        }
+        System.out.println(newText3);
```
Comment (TheFallen058) [inline src/lesson5/javaBeginnerChapter2/Task21.java:57]: Did you analyze results? What is more efficient?

File: src/lesson5/javaBeginnerChapter2/Task21.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                }
+            }
+            long end = System.nanoTime();
+            System.out.println("String: " + newText2);
+            System.out.println("Total execution time: " + (end - start));
+        }
+        System.out.println(newText2);
+
+        System.out.print("Enter string(class StringBuffer): ");
+        StringBuffer str3 = new StringBuffer(sc.nextLine());
```
Comment (TheFallen058) [inline src/lesson5/javaBeginnerChapter2/Task21.java:44]: What is difference between string builder and string buffer? Why does IDEA suggest in this place to use string builder?

**Review by TheFallen058 — APPROVED** [review]
Great work! Please just check my comments

File: src/lesson5/javaBeginnerChapter2/Task21.java (hunk 3 above)
Comment (Pokosteler) [reply src/lesson5/javaBeginnerChapter2/Task21.java:57]: The third result is more effective than the others

File: src/lesson5/javaBeginnerChapter2/Task21.java (hunk 4 above)
Comment (Pokosteler) [reply src/lesson5/javaBeginnerChapter2/Task21.java:44]: StringBuffer is synchronized i.e. thread safe. StringBuilder is non-synchronized i.e. not thread safe.

Comment (TheFallen058) [reply src/lesson5/javaBeginnerChapter2/Task21.java:44]: It's just words from the book) Our current logic doesn't contain multithreading, so no sense to use StringBuffer (IDEA shows it in the hint). All synchronized methods and collections will work same or longer than their non-synchronized analogs. So it we compare StringBuilder and StringBuffer - first one is more efficient, but we can't use it in multithreading constructions and should use StringBuffer.
Let's finalize: 
If we create single string - use new String() constructor.
If we create string in cycle - StringBuilder
if it's cycle in multiple threads - StringBuffer

File: src/lesson5/javaBeginnerChapter2/Task21.java (hunk 3 above)
Comment (TheFallen058) [reply src/lesson5/javaBeginnerChapter2/Task21.java:57]: please run it at list 3 times (better 10) and compare results again.

Comment (Pokosteler) [reply src/lesson5/javaBeginnerChapter2/Task21.java:57]: After 3 or 4 launches, StringBuilder was more effective

**Pull request merged**